git/odb/object_writer.go

```diff
@@ -0,0 +1,123 @@
+package odb
+
+import (
+	"compress/zlib"
+	"crypto/sha1"
+	"fmt"
+	"hash"
+	"io"
+	"sync/atomic"
+)
+
+// ObjectWriter provides an implementation of io.Writer that compresses and
+// writes data given to it, and keeps track of the SHA1 hash of the data as it
+// is written.
+type ObjectWriter struct {
+	// w is the underling writer that this ObjectWriter is writing to.
+	w io.Writer
+	// sum is the in-progress hash calculation.
+	sum hash.Hash
+
+	// wroteHeader is a uint32 managed by the sync/atomic package. It is 1
+	// if the header was written, and 0 otherwise.
+	wroteHeader uint32
+
+	// closeFn supplies an optional function that, when called, frees an
+	// resources (open files, memory, etc) held by this instance of the
+	// *ObjectWriter.
+	//
+	// closeFn returns any error encountered when closing/freeing resources
+	// held.
+	//
+	// It is allowed to be nil.
+	closeFn func() error
+}
+
+var _ io.WriteCloser = (*ObjectWriter)(nil)
+
+// nopCloser provides a no-op implementation of the io.WriteCloser interface by
+// taking an io.Writer and wrapping it with a Close() method that returns nil.
+type nopCloser struct {
+	// Writer is an embedded io.Writer that receives the Write() method
+	// call.
+	io.Writer
+}
+
+var _ io.WriteCloser = (*nopCloser)(nil)
+
+// Close implements the io.Closer interface by returning nil.
+func (n *nopCloser) Close() error {
+	return nil
+}
+
+// NewObjectWriter returns a new *ObjectWriter instance that drains incoming
+// writes into the io.Writer given, "w".
+func NewObjectWriter(w io.Writer) *ObjectWriter {
+	return NewObjectWriteCloser(&nopCloser{w})
+}
+
+// NewObjectWriter returns a new *ObjectWriter instance that drains incoming
+// writes into the io.Writer given, "w".
+//
+// Upon closing, it calls the given Close() function of the io.WriteCloser.
+func NewObjectWriteCloser(w io.WriteCloser) *ObjectWriter {
```

Comment:
@technoweenie good catch. I think having these be called `NewObjectReader` and `NewObjectReadCloser` (with identical `Writer` equivalents) is preferable to dropping "Object" from the signature, and I've gone ahead and made that change in 2c6ddd6c.

I think keeping "Object" is important, because these are not Object Database readers (as odb.Reader` would suggest). Rather, they are Object Readers, reading individual objects, not the whole database.